### ISSUE TITLE
allow build without gradle

### ISSUE THREAD

**Issue by judovana** [policeman-tools/forbidden-apis#158]
To exclude gradle itself is moreover easy. But that removes groovy.  WDYT?

**Comment by uschindler** [comment]
What's exactly the problem?

**Comment by judovana** [comment]
Hi!  The patches to build without gradle looks like this:
https://src.fedoraproject.org/rpms/forbidden-apis/blob/master/f/fix-gradle-maven-build.patch
+ https://src.fedoraproject.org/rpms/forbidden-apis/blob/master/f/forbidden-apis.spec#_72

I'm not sure what the groovy script is dooing, and whether forbidden apis are still doing what they should (I built it for gradle-less fedora just now).

If I drop gradle, I have to drop groovy.  Anyway, I was more thinking about build switch of no_gradle.

Thanx for promt reply!

**Comment by uschindler** [comment]
Hi judovana (are you Jiri?),
no problem, thanks for the clarification. I was expecting that a linux distribution was behind this request :-)

I will carefully answer your question:

> I'm not sure what the groovy script is dooing, and whether forbidden apis are still doing what they should (I built it for gradle-less fedora just now).

The Groovy script is not necessary to actually build everything from source. As you might have seen from the patch, it's part of a custom task. IMHO, in the patch, you should remove the generate-* tasks completely, because those depend on Groovy. But actually for a Linux distribution (I refer here to Debian, who are very paranoid with external stuff) it might still be good to make sure that the generate-* tasks are applied. 

Why are the generate-tasks not needed? They are used to produce some of the signatures files (main ly the deprecated signatures). They are not hand-maintained, but generated from the JDK class library. As you need to run the code to generate them on all and every supported JDK (6, 7, 8, 9, 10, 11, 12, 13 - soon 14) that's a big hassle: https://github.com/policeman-tools/forbidden-apis/tree/master/src/main/resources/de/thetaphi/forbiddenapis/signatures; so they are one time generated after release of a new JDK and before a new release of forbiddenapis. Some of the files in the above folder have no license header, because they are autogenerated (`jdk-deprecated-*` and `jdk-internal-*`). From our checks and communication with the OpenJDK / Oracle guys, this is no problem, but if you are paranoid you would need to regenerate them. To regenerate, the ant tasks `ant generate-deprecated` and `ant generate-internal` are used as part of code maintenance. If you don't delete above files, you can remove groovy.

> If I drop gradle, I have to drop groovy. Anyway, I was more thinking about build switch of no_gradle.

This should be unrelated to each other. You can remove gradle and leave groovy.

Now some general comments:

- I figured out in the above patch, that you seem to patch the Gradle Task Java file, but the spec file deletes it afterwards (the whole gradle folder is removed from source). I was wondering what the removal/patching of the annotation helps, the whole file relies on Gradle. :-)
- As noted before, I'd suggest to remove the `generate-deprecated` ANT task completely if you remove Groovy.
- I don't like to make Gradle support to be compiled "configurable". The reason for this is backed by a general problem that most Java / Maven /Gradle plugin developers have with Linux distributions: Linux distributions mostly try to be self contained, they rebuild everything from source (instead of using Maven central). From my background with Apache Lucene, I know how horrible that is, because the build environments are just patched and the JAR files produced don't behave identical to the original ones, often be more buggy and lacking fetaures. People relying on them can never be sure that everything works as expected (it's like here: You don't know if everything works). With Lucene we had the problem that Debian rebuilds JavaCC files, but uses wrong JDK for that, which leads to different text analysis (because Unicode versions of newer JDKs change), breaking people's index. Another issue is also related to build tools: The default ANT as shipped with Fedora is unuseable to build e.g., Lucene, because it pollutes classpath with unrelated libraries (as does forbiddenapis after your patch). ASM is very problematic, e.g. if somebody installs forbiddenapis in the Fedora system, it may be installed in some ANT/Maven Lib folder and because ASM is not shaded/jarjared, it will break other tools.

In short: the official way to use forbiddenapis is by referring to it in your ivy/maven/gradle file and suck it from Maven Central. If we support different JAR file variants with reduced behaviour, how would somebody e.g. build Elasticsearch on his Fedora system when using the native forbiddenapis shipped with operating system? So please ship the JAR file in unmodified form and not stripped down functionality. You can build it from source, but please don't modify features. What's the problem with building the Gradle plugin correctly?

I hope the previous info helps to understand the position.

**Comment by judovana** [comment]
> Hi judovana (are you Jiri?),

Yup!  Sorry for the confusion.

> no problem, thanks for the clarification. I was expecting that a linux distribution was behind this request :-)
> 
> I will carefully answer your question:
> 
> > I'm not sure what the groovy script is dooing, and whether forbidden apis are still doing what they should (I built it for gradle-less fedora just now).
> 
> The Groovy script is not necessary to actually build everything from source. As you might have seen from the patch, it's part of a custom task. IMHO, in the patch, you should remove the generate-* tasks completely, because those depend on Groovy. But actually for a Linux distribution (I refer here to Debian, who are very paranoid with external stuff) it might still be good to make sure that the `generate-*` tasks are applied.
> 

I see. Groovy in fedora was depending on gradle, so it went out with it. So you are right, you do not need gradle for grovy in runtime, but you need it in build time. Once those are connected, yo suddenly needs gradle for groovy:(

> Why are the `generate`-tasks not needed? They are used to produce some of the signatures files (main ly the deprecated signatures). They are not hand-maintained, but generated from the JDK class library. As you need to run the code to generate them on all and every supported JDK (6, 7, 8, 9, 10, 11, 12, 13 - soon 14) that's a big hassle: https://github.com/policeman-tools/forbidden-apis/tree/master/src/main/resources/de/thetaphi/forbiddenapis/signatures; so they are one time generated after release of a new JDK and before a new release of forbiddenapis. Some of the files in the above folder have no license header, because they are autogenerated (`jdk-deprecated-*` and `jdk-internal-*`). From our checks and communication with the OpenJDK / Oracle guys, this is no problem, but if you are paranoid, you would need to regenerate them (downloading and installing temporary each of the above JDKs). To regenerate, the ant tasks `ant generate-deprecated` and `ant generate-internal` are used as part of code maintenance. If you don't delete above signatures files, you can remove groovy.
> 

Thanx a lot for explanation! That explains why I had hard times to find the connection. I will happily keep the pre-generated signature files.

> The reason why it's written in Groovy is as its dynamically compiled. The script works with all JDK versions, where a compiled version would not even build with older JDKS, leading to an chicken-egg problem.

Sure.  I like groovy:)  And this is nice example of its usage.
> 
> > If I drop gradle, I have to drop groovy. Anyway, I was more thinking about build switch of no_gradle.
> 
> This should be unrelated to each other. You can remove gradle and leave groovy.

Right, except the build chahin as told above:( But with the expalined usage.. no more needs to look ito it. Really thanx!
> 
> Now some general comments:
> 
>     * I figured out in the above patch, that you seem to patch the Gradle Task Java file, but the spec file deletes it afterwards (the whole gradle folder is removed from source). I was wondering what the removal/patching of the annotation helps, the whole file relies on Gradle. :-)
> 

Yah.  noticed it right after push in above shining diff.  I will remove it in next  iteration. At least it amde your day a bit more funnier :)

>     * As noted before, I'd suggest to remove the `generate-deprecated` ANT task completely if you remove Groovy.

Will keep it in mind.  A  bt more is now broken in fedora javastack (mostly exactly because of the reasons yu descibe lower)
> 
>     * I don't like to make Gradle support to be compiled "configurable". The reason for this is backed by a general problem that most Java / Maven /Gradle plugin developers have with Linux distributions: Linux distributions mostly try to be self contained, they rebuild everything from source (instead of using Maven central). From my background with Apache Lucene, I know how horrible that is, because the build environments are just patched and the JAR files produced don't behave identical to the original ones, often be more buggy and lacking fetaures. People relying on them can never be sure that everything works as expected (it's like here: You don't know if everything works). With Lucene we had the problem that Debian rebuilds JavaCC files, but uses wrong JDK for that, which leads to different text analysis (because Unicode versions of newer JDKs change), breaking people's index. Another issue is also related to build tools: The default ANT as shipped with Fedora is unuseable to build e.g., Lucene, because it pollutes classpath with unrelated libraries (as does forbiddenapis after your patch). ASM is very problematic, e.g. if somebody installs forbiddenapis in the Fedora system, it may be installed in some ANT/Maven Lib folder and because ASM is not shaded/jarjared, it will break other tools.
> 
> 
> In short: the official way to use forbiddenapis is by referring to it in your ivy/maven/gradle file and suck it from Maven Central. If we support different JAR file variants with reduced behaviour, how would somebody e.g. build Elasticsearch on his Fedora system when using the native forbiddenapis shipped with operating system? So please ship the JAR file in unmodified form and not stripped down functionality. You can build it from source, but please don't modify features. What's the problem with building the Gradle plugin correctly?
> 

Fedora dropped gradle. That is all. But quite crucial :) So the idea is to build maven and ivy plugins, so fedora toolchain can continue.
But I understand your ponit of view. So definitely closing as upstream issue., as it remains downstream, even fedora, only. Thanx a lot for hints how to do it in better way.
Me as developer, relay on centrals. Me as linux user relyes on repositories. Those are hard decissions from time to time.

> I hope the previous info helps to understand the position.

hank you very much! Your advices were direct and helpful.
  J.
